File: demo-android/app/src/main/java/com/example/demo/ui/i18n/ErrorMessages.kt
package com.example.demo.ui.i18n

import android.content.Context
import com.example.demo.R
import com.example.demo.data.remote.ApiClient
import com.example.demo.locale.AppLocale
import retrofit2.HttpException

fun mapTaskError(context: Context, throwable: Throwable): String {
    if (throwable is HttpException) {
        val message = ApiClient.parseErrorMessage(
            throwable.response()?.errorBody()?.string()
        )
        if (!message.isNullOrBlank()) {
            return message
        }
        return when (throwable.code()) {
            400 -> AppLocale.getString(context, R.string.error_invalid_task_data)
            404 -> AppLocale.getString(context, R.string.error_task_not_found)
            409 -> AppLocale.getString(context, R.string.error_title_already_exists)
            else -> AppLocale.getString(
                context,
                R.string.error_request_failed,
                throwable.code()
            )
        }
    }
    return throwable.message ?: AppLocale.getString(context, R.string.error_something_went_wrong)
}

fun isDuplicateTitleError(throwable: Throwable): Boolean =
    throwable is HttpException && throwable.code() == 409

File: demo-react-native/src/i18n/errorMessages.ts
import {TFunction} from 'i18next';

import {ApiError} from '@/data/remote/apiClient';

export function mapTaskError(t: TFunction, error: unknown): string {
  if (error instanceof ApiError) {
    if (error.message && !error.message.startsWith('Request failed')) {
      return error.message;
    }
    switch (error.status) {
      case 400:
        return t('errorInvalidTaskData');
      case 404:
        return t('errorTaskNotFound');
      case 409:
        return t('errorTitleAlreadyExists');
      default:
        return t('errorRequestFailed', { code: error.status });
    }
  }
  if (error instanceof Error && error.message) {
    return error.message;
  }
  return t('errorSomethingWentWrong');
}

export function isDuplicateTitleError(error: unknown): boolean {
  return error instanceof ApiError && error.status === 409;
}
